packages/vscode-extension/src/utilities/throttle.ts

```diff
@@ -28,3 +28,44 @@ export function throttle<T extends AnyFn>(func: T, limitMs: number): WithForce<T
     recentArgs = args;
   } as T;
 }
+
+type AsyncFn = (...args: any[]) => Promise<any>;
+
+/**
+ * Throttles the provided async function with the following restrictions:
+ * 1) Function will never be called more than once every `limitMs` milliseconds.
+ * 2) Function will never be called concurrently.
+ * 3) If called multiple times, it is guaranteed that the last call with the final
+ *   arguments will be executed.
+ */
+export function throttleAsync<T extends AsyncFn>(func: T, limitMs: number): T {
+  let timeout: NodeJS.Timeout | null = null;
+  let recentArgs: any;
+
+  return async function (...args: any) {
+    if (!timeout) {
+      const execute = () => {
+        const currentArgs = recentArgs;
+        const result = func(...recentArgs);
+        result
+          .catch(() => {})
+          .then(() => {
+            if (recentArgs === currentArgs) {
+              timeout = null;
+              recentArgs = null;
+            } else {
+              // if new arguments were provided while the function was executing,
+              // we need to run it again to ensure the last call is executed.
+              // we use setTimeout to avoid potentially infinite recursion, and
+              // to ensure the function is not run more often than once every
+              // `limitMs` milliseconds, we schedule it to run after the provided
+              // limit.
+              setTimeout(execute, limitMs);
```

Comment:
nit: if you are not using the timeout to cancel it, but just to check if it was completed or not it might be better to keep a simple boolean value instead, but if you want to keep the current timeout in the value `timeout` you should assign it here as well

Reply:
that's a good catch too. I copied it from regular throttle where we have "force" method that'd cancel the timeout. This isn't needed here though